app/controllers/Checkout.scala

```diff
@@ -16,25 +17,32 @@ object Checkout extends Controller {
       IdentityService.userLookupByScGuU(cookie.value)
     }.getOrElse(Future.successful(None))
 
-    idUserFutureOpt.map { idUserOpt =>
+    for (idUserOpt <- idUserFutureOpt) yield {
       def idUserData(fieldName: PrivateFields => Option[String]): Option[String] =
         idUserOpt.flatMap(_.privateFields.flatMap(fieldName))
 
-      Ok(
-        views.html.checkout.payment(
-          idUserData(_.firstName),
-          idUserData(_.secondName),
-          idUserOpt.map(_.primaryEmailAddress),
-          idUserData(_.billingAddress1),
-          idUserData(_.billingAddress2),
-          idUserData(_.billingAddress3),
-          idUserData(_.billingPostcode))
+      val form = SubscriptionsForm().copy(
+        data =  (
```

Comment:
Wow! That `Option.++` is clever, I didn't know it existed!

I think this code could be even more compact if we were to produce the tuples directly in the `idUserData` function:

```scala
      def idUserData(keyName: String, fieldName: PrivateFields => Option[String]): Option[(String, String)] =
        for {
          idUser <- idUserOpt
          fields <- idUser.privateFields
          field  <- fieldName(fields)
        } yield keyName -> field

      val form = SubscriptionsForm().copy(
        data =  (
          idUserData("personal.first", _.firstName)                                  ++
          idUserData("personal.last", _.secondName)                                  ++
          idUserOpt.map("personal.emailValidation.email" -> _.primaryEmailAddress)   ++
          idUserOpt.map("personal.emailValidation.confirm" -> _.primaryEmailAddress) ++
          idUserData("personal.address.address1", _.address1)                        ++
          idUserData("personal.address.address2", _.address2)                        ++
          idUserData("personal.address.town", _.address3)                            ++
          idUserData("personal.address.postcode", _.postcode)
        ).toMap
      )
```